### ISSUE TITLE
Native re-implementation of Web Discovery (part 1)

### ISSUE THREAD

**Issue by DJAndries** [brave/brave-browser#39439]
Currently, the Web Discovery Project exists as part of the embedded Brave extension. In order for Android/iOS users to opt-in and participate in Web Discovery, the client must be re-implemented natively.

Web Discovery sends three types of payloads:

- `alive`: a ping to indicate to the Web Discovery servers that the user is opted in
- `query`: search engine results, if the query is deemed to not contain private information
- `page`: page interaction events, to measure user engagement for non-private pages

The initial re-implementation will only cover the first two payloads. The third payload will be covered in a separate issue. In addition, a separate issue will be needed for iOS in order to cover some WebKit renderer necessities.

A feature controlled by variations (`BraveWebDiscoveryNative`) will be used to handle the rollout to mobile users. Once the `page` payload is implemented, we can begin to deprecate the extension code by rolling the feature out to desktop users.

cc @anthonypkeane @rebron @bbondy @aekeus @remusao

**Comment by LaurenWags** [comment]
@DJAndries is this ready for QA? If so, can you please add a test plan? Thanks!

**Comment by DJAndries** [comment]
> [@DJAndries](https://github.com/DJAndries) is this ready for QA? If so, can you please add a test plan? Thanks!

Hey @LaurenWags , we are still waiting for the Android scraping rules to be published. Once that is complete (hopefully over the next couple of days), I will add a test plan, thank you.

cc @remusao

**Comment by LaurenWags** [comment]
@DJAndries can we get an update on this please? Thanks.

cc @rebron @kjozwiak @bsclifton

**Comment by DJAndries** [comment]
> [@DJAndries](https://github.com/DJAndries) can we get an update on this please? Thanks.
> 
> cc [@rebron](https://github.com/rebron) [@kjozwiak](https://github.com/kjozwiak) [@bsclifton](https://github.com/bsclifton)

The Search team are still in the process of updating their scraping patterns, but they confirmed that they should have it wrapped up by the end of the week. I hope to have a test plan ready tomorrow or early next week.

**Comment by DJAndries** [comment]
Here is a test plan, please hold until https://github.com/brave/brave-core/pull/27788 is merged and uplifted:

0. Start with fresh profile, configure a MITM proxy
1. Proceed with onboarding, do NOT enable WDP when prompted
2. Enter QA settings, enter the following command line string: `--enable-logging=stderr --vmodule="*/web_discovery/*=2" --wdp-collector-host=https://collector.wdp.brave.software --enable-features="BraveWebDiscoveryNative" --wdp-patterns-url=https://djandries.github.io/patterns.gz`
3. Upon restart, wait a minute, ensure that there are no requests to `collector.wdp.brave.software`, `quorum.wdp.brave.com` or `djandries.github.io` logged in the proxy
4. Go to privacy settings, enable Web Discovery
5. Ensure three `/join` requests are sent to `collector.wdp.brave.software`, each with a HTTP 200 response.
6. Ensure a request is sent to `https://quorum.wdp.brave.com`
7. Wait a minute, ensure a request to `https://djandries.github.io/patterns.gz` is made
8. Go to google.com, query "best cookie recipes"
9. Watch adb for relevant logs. Ensure the following log appears: `Double fetching search page: https://www.google.com/search?q=best+cookie+recipe`
10. Open a new tab, close the Google tab.
11. Wait a minute. Ensure two requests to `https://www.google.com/search?q=best+cookie+recipes` are sent around a minute after the original request.
12. Wait another minute, ensure another request is sent to `https://collector.wdp.brave.software/` with a HTTP 200 response.
13. Look for a log prefixed with "Preparing to report payload". Ensure that the payload structure and content is similar to payload provided below these steps.
14. Ensure a request was sent to `POST https://collector.wdp.brave.software/` with a HTTP 200 response. Ensure a `Submission result: 1` log appears below the "preparing" request.
15. Make another query for "best cookie recipes". Wait a few minutes. Ensure a request is NOT sent to `POST https://collector.wdp.brave.software/`.
16. Make another query for "best cake recipes". Wait a few minutes. Ensure a request IS sent to `POST https://collector.wdp.brave.software/`. Ensure that there is another "preparing" log with contents relevant to the query.

At some point, a "preparing" log with the `action` field set to `alive` will be sent. There will be a request sent to `POST https://collector.wdp.brave.software/` to accompany this log.

**Comment by LaurenWags** [comment]
Thanks @DJAndries - looks like this test plan is Android specific though. What should be checked for desktop since this issue has `OS/Desktop` label? Or are checks on desktop not needed?

cc @kjozwiak

**Comment by DJAndries** [comment]
> Thanks [@DJAndries](https://github.com/DJAndries) - looks like this test plan is Android specific though. What should be checked for desktop since this issue has `OS/Desktop` label? Or are checks on desktop not needed?
> 
> cc [@kjozwiak](https://github.com/kjozwiak)

We won't be using the native implementation on Desktop just yet. I went ahead and removed the `OS/Desktop` label; no QA is needed for Desktop.

**Comment by kjozwiak** [comment]
So https://github.com/brave/brave-core/pull/27788 was merged into `master` and uplifted into `1.76.x` via https://github.com/brave/brave-core/pull/27806. However, not removing the `QA/Blocked` as @DJAndries mentioned we need one more issue/uplift before being able to run through https://github.com/brave/brave-browser/issues/39439#issuecomment-2680215238. @DJAndries mind listing/mentioning the PR/issue so we can keep an eye on it?

**Comment by DJAndries** [comment]
QA is blocked until https://github.com/brave/brave-core/pull/27814 is merged

**Comment by kjozwiak** [comment]
> QA is blocked until [brave/brave-core#27814](https://github.com/brave/brave-core/pull/27814) is merged

@Uni-verse @hffvld unblocked as the above was merged/uplifted into `1.76.x`. Requires `1.76.70` or higher for https://github.com/brave/brave-browser/issues/39439#issuecomment-2680215238 to be verified.

**Comment by hffvld** [comment]
Verified on `Galaxy Tab S8` and `Pixel 7` using version(s):
```
Device/OS: 
- Galaxy Tab S8 / gts8wifixx-user 14 UP1A.231005.007 release-keys
- Pixel 7 / panther_beta-user 16 BP22.250103.008 release-keys
Brave build: 1.76.70 
Chromium: 134.0.6998.39 (Official Build) (64-bit) 
```
---------------------------

**STEPS:**
1. Follow the STR/TP from https://github.com/brave/brave-browser/issues/39439#issuecomment-2680215238
2. *Verify*

**ACTUAL RESULTS:**
- Verified the above behavior is correctly reproduced.

<details>
<summary> Pixel 7 </summary>

---------------------------

| Web Discovery is off | Web Discovery is on: `/join` request 1 | Web Discovery is on: `/join` request 2 | Web Discovery is on: `/join` request 3 |
|:-:|:-:|:-:|:-:|
| ![Image](https://github.com/user-attachments/assets/7056f6d7-7d05-44df-8c69-dbbc446e969f) | ![Image](https://github.com/user-attachments/assets/7eef8ade-cd9f-447c-84fe-f74e60964f76) | ![Image](https://github.com/user-attachments/assets/7918583d-7a65-4bdb-98a2-8b08453ed132) | ![Image](https://github.com/user-attachments/assets/50653cd4-4a4e-40ed-9e2b-d09b09e96efb) |
| `https://quorum.wdp.brave.com` request | `https://djandries.github.io/patterns.gz` request | `https://www.google.com/search?q=best+cookie+recipes` original request 1 | `https://www.google.com/search?q=best+cookie+recipes` original request 2 |
| ![Image](https://github.com/user-attachments/assets/ba643892-967f-4648-bff1-5da07c5c1494) | ![Image](https://github.com/user-attachments/assets/5a37926d-8b20-4062-86fa-cce375b7ff15) | ![Image](https://github.com/user-attachments/assets/25756400-31e2-4a73-8fb9-2ee9d7717bbb) | ![Image](https://github.com/user-attachments/assets/7c02d344-8b3f-4a7f-bf67-fbaa66fb73e3) |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |




</details>
